Connect programs report page to the API data.

tomalec opened the pull request:
### Changes proposed in this Pull Request:

- Add `useProgramsReport` hook, with glue code for aggregating free and paid programs. (b79c9fc)
- Add mocked `(ads|mc)/reports/programs` responses for testing (075cd8d)
- Connect programs `SummarySection` to API data (74b0308)
	- Add optional `expectedLength` prop to `SummarySection`, so it could render applicable placeholder even before the metrics array is loaded/resolved.
- Add and connect programs's `ChartSection` to the API,(695f91d)
	- Make `ChartSection` accept empty `metrics` prop, so it could render a placeholder, when `metrics` data is being loaded/resolved.


Implements part of #242 

### Screenshots:

![Screencast of clicking through program reports connected to data](https://user-images.githubusercontent.com/17435/118124444-e0836900-b3f5-11eb-929a-dd38517cad5e.gif)



### Detailed test instructions:

1. Connect to an account that has some reports data, or use mocked responses, by adding at the end of [`google-listings-and-ads.php`](https://github.com/woocommerce/google-listings-and-ads/blob/trunk/google-listings-and-ads.php#L91) following snippet:
	  ```php
	  
	  add_filter(
		  'gla_prepared_response_ads-reports-programs',
		  function( $response ) {
			  return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
		  },
	  );
	  add_filter(
		  'gla_prepared_response_mc-reports-programs',
		  function( $response ) {
			  return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
		  },
	  );
	  ```
2. Go to Programs Report page  [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms)
3. Click through Summary Numbers and check that Chart updates accordingly,
4. Check that tooltip about missing data is rendered,
5. Switch programs filter value to `Single program->"Free Listings"`
6. Redo 3 & 4
7. Switch programs filter value to `Single program->any "Campaign"`
6. Redo 3 & 4

### Not covered in this PR:
1. Intervals data for the chart is not aggregated, what results with chart line going back https://github.com/woocommerce/google-listings-and-ads/issues/589 
2. Programs Table will be updated in another PR

### Doubts:

1. Currently, until reports data come we render a fancy preloader for all 5 metrics, which may eventually be replaced with only 2 metrics, if 
	1. User selected `Single program -> "Free Listings"`
	2. User selected `All` and does not have any paid campaigns
	3. User selected `All` but paid campaigns do not have any report data yet.
	4. User selected `Compare`, and choose `free listings, campaign#12` but the paid campaign no longer exists.

	I'm not sure if that 100% desired behavior. The alternative solution that comes to my mind is:
	Decide the number of Metrics based on the selection, not the actually available data, and show 
	_i._ - a placeholder for 2 metrics that resolves to 2 metrics with data,
	_ii., iii., iv._ - a placeholder with 5 metrics that resolves to 5 metrics but some with `0` or `N/A`.
	

### Changelog Note:

> Connected Programs Report Summary and Chart to API data.

tomalec committed:
Move `programs` query param name to contants.

tomalec committed:
Add `useProgramsReport` hook, with glue code for aggregating free and paid programs.

tomalec committed:
Add mocked `(ads|mc)/reports/programs` responses for testing

tomalec committed:
Connect programs `SummarySection` to API data,

Add optional `expectedLength` prop to `SummarySection`, soit could render applicable placeholder even before the metrics array is loaded/resolved.

tomalec committed:
Connect programs report's `ChartSection` to the API,

use reused component.
Make `ChartSection` acceptempty `metrics` prop, so it could render a placeholder, when `metrics` data is being loaded/resolved.

tomalec committed:
Remove unused, legacy mocked metrics data.

eason9487 reviewed (commented):
Other issues or suggestions may be addressed with this PR or later
-  💅 . The metric title is still "Net Sales"
    ![image](https://user-images.githubusercontent.com/17420811/118240749-ecdaf500-b4cd-11eb-9c25-5f44cb9ac1f9.png)
- 🚧 . `selectedMetric` needs to be one of `staticParams` item in [here](https://github.com/woocommerce/google-listings-and-ads/blob/83a9005c74407ef77fcc3061e31f91b29713f612/js/src/reports/programs/filter-config.js#L55) to make sure the `selectedMetric` in the URL query will not be removed after changed the **Show** filter. Reproduce steps:
    1. Select the "Impressions" metric by summary section. Then the `selectedMetric` is appended onto the URL query
    2. Change "All Google programs" to any single program by the **Show** filter option
    3. The `selectedMetric` is removed
- 🚧 . Render break after changing **Show** option. (I used a [workaround](https://github.com/woocommerce/google-listings-and-ads/blob/83a9005c74407ef77fcc3061e31f91b29713f612/js/src/reports/products/filter-config.js#L27-L32) to avoid it)
    1. Enter the programs report page by default URL path and make you have at least one paid grogram
    2. Select the "Conversions" metric by summary section
    3. Change **Show** option from "All Google programs" to "Single program" and pick "Free Listings"
    ![Kapture 2021-05-14 at 16 27 08](https://user-images.githubusercontent.com/17420811/118243917-9079d480-b4d1-11eb-907d-617300a9c2bf.gif)

tomalec committed:
Fix JSDoc descriptions for Reports,

Apply suggestions from https://github.com/woocommerce/google-listings-and-ads/pull/602#pullrequestreview-659559236

Co-authored-by: Eason <eason.su.tw@gmail.com>

tomalec committed:
Merge branch 'trunk' into feature/242-program-reports-data

tomalec committed:
Use shared `getIdsFromQuery` util in `useProgramsReport`,

Follow-up after https://github.com/woocommerce/google-listings-and-ads/pull/606.

tomalec committed:
Set default value in `SummarySection` props destructor.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/602#discussion_r632326697

tomalec committed:
Fix JSDoc types for ProgramsReportSchema.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/602#discussion_r632342827

tomalec committed:
Fix the paid + free aggregation,

Correct the arguments order, address https://github.com/woocommerce/google-listings-and-ads/pull/602#discussion_r632352895
Consider `undefined` & any falsy metric value as `0`, address https://github.com/woocommerce/google-listings-and-ads/pull/602#discussion_r632356080

tomalec committed:
Rename `Net Sales` to `Total Sales` in Programs Report

Address https://github.com/woocommerce/google-listings-and-ads/pull/602#pullrequestreview-659559236

tomalec committed:
Preserve selectedMetric upon filtering programs.

Address https://github.com/woocommerce/google-listings-and-ads/pull/602#pullrequestreview-659559236

tomalec committed:
Fix render error, when unavailable `selectedMetric` is given to `ChartSection`

Address https://github.com/woocommerce/google-listings-and-ads/pull/602#pullrequestreview-659559236

eason9487 reviewed (approved):
Tested locally and works well, LGTM. Left a comment for polishing a little of JSDoc.

tomalec committed:
Fix JSDocs for `useProgramsReport`,

Switch `products` with `programs`,
addresses:
https://github.com/woocommerce/google-listings-and-ads/pull/602#pullrequestreview-660519130
